### ISSUE TITLE
fallback to loading packages by name

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/Pkg.jl#457]
In general, I'm finding package-resolution failures to be a huge pain during the 0.7 upgrade process (e.g. #367).

Couldn't the module loading fall back to loading by name and only printing a warning (giving an error only if there are multiple installed package UUIDs with the same name)?   A warning should be sufficient to get people to try to specify the correct UUID eventually, but doesn't force you to do it immediately.

**Comment by KristofferC** [comment]
I have asked a few times if falling back to the Project UUID for packages that are currently developed is a reasonable trade-off. Opinions?

**Comment by stevengj** [comment]
Yes, with a warning.

**Comment by StefanKarpinski** [comment]
So what would the rule be? If you try to load a name that’s in the manifest and unique but not properly in the deps then load it with a warning?

**Comment by KristofferC** [comment]
Yes, thats what I had in mind.

**Comment by KristofferC** [comment]
> If I just do Pkg.dev("Foo"), then the "Project UUID" for a using Bar inside Foo.jl is the same as the UUID that is used to resolve using Bar at the REPL, right?)

In the REPL it would be determined from the  Project.toml entry with `Bar = "&Uuid"`.

In `Foo` it is determined by the Manifest.toml in a `deps = ` section under the `Foo` section. If `Foo` is unique then it would just be stored as `deps = [Foo] ` otherwise it would also contain the uuid of `Foo` there. That is because `Foo` can use a package called `Bar` which is different from the "top level" one.

**Comment by stevengj** [comment]
> That is because Foo can use a package called Bar which is different from the "top level" one.

Yes, I understand that, but for the case where `Bar` is not properly listed in the deps, the proposed behavior is that it would use the top-level `Bar` if possible (with a warning), correct?

**Comment by KristofferC** [comment]
Yes.

**Comment by simonbyrne** [comment]
Is this what is causing the recent Travis failures? e.g.
```
ERROR: The following package names could not be resolved:
 * Arpack (not found in project or manifest)
Please specify by known `name=uuid`.
Stacktrace:
 [1] #ensure_resolved#29(::Bool, ::Function, ::Pkg.Types.EnvCache, ::Array{Pkg.Types.PackageSpec,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Pkg/src/Types.jl:119
 [2] ensure_resolved at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Pkg/src/Types.jl:729 [inlined]
 [3] #build#51(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Pkg/src/API.jl:395
 [4] build at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Pkg/src/API.jl:377 [inlined]
 [5] build(::String) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Pkg/src/API.jl:371
 [6] top-level scope at none:0
```
from https://travis-ci.org/JuliaLinearAlgebra/Arpack.jl/jobs/399348800

**Comment by KristofferC** [comment]
No, I think that is the new code loading that requires a `--project` flag to the executable (alt `JULIA_PROJECT=@.`)  to start the session with the current dir as the active project.

**Comment by stevengj** [comment]
Fixed by JuliaLang/julia#27932
